feat: enable logging for `defineData` (#8190)
aws-amplify/docs: src/pages/[platform]/build-a-backend/data/enable-logging/index.mdx

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {

>   title: 'Enable logging for your data resource',
>   description: 'Enable logging for your data resource',

Reviewer: ```suggestion
  title: 'Enable logging',
  description: 'Learn how to enable logging for your Amplify data resource',
```

  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

> You can enable logging to debug your GraphQL API using CloudWatch logs.

Reviewer: ```suggestion
You can enable logging to debug your GraphQL API using Amazon CloudWatch logs.
```

the first reference of a service name on a page should use the full service name

> ## Default Logging

Reviewer: ```suggestion
## Enable default logging configuration
```

"Default logging" is a bit ambiguous where we're wanting to showcase the default logging configuration

Default logging can be enabled by setting the `logging` property to `true` in the call to `defineData`. For example:

```ts title="amplify/data/resource.ts"
export const data = defineData({

>   some_schema,
>   some_authorizationModes,

Reviewer: ```suggestion
  // ...
```

I think for this we can either show a full example with stubs for the property values or a comment that denotes some information precedes the logging prop. This example is using the property shorthand syntax assigning the variable `some_schema` to the property `some_schema`, i.e.

```ts
const some_schema = a.schema({})

defineData({
  some_schema: some_schema,
})
```

  logging: true
});
```

> Using `logging: true` applies the default configuration:

Reviewer: can we add a link here to the property descriptions towards the bottom of the page or link out to an explainer? as a customer reading this page for the first time, `excludeVerboseContent` is an ambiguous term

Author: I added the links for all 3 of them to keep it consistent. These links are repeated again below in Configuration Properties section

- `excludeVerboseContent: true`
- `fieldLogLevel: 'none'`
- `retention: '1 week'`

> ## Custom Logging

Reviewer: ```suggestion
## Customize logging configuration
```

Reviewer: similar suggestion as previous heading

> You can override individual fields by providing a `DataLogConfig` object. For example:

Reviewer: ```suggestion
You can customize individual configuration values by providing a [`DataLogConfig`](#datalogconfig-fields) object. For example:
```

nit for the term "override" since we typically reserve this for overriding resource properties via CDK

```ts title="amplify/data/resource.ts"
export const data = defineData({

>   some_schema,
>   some_authorizationModes,

Reviewer: ```suggestion
  // ...
```

similar callout as above for the shorthand property syntax

  logging: {
    excludeVerboseContent: false,
    fieldLogLevel: 'all',
    retention: '1 month'
  }
});
```

<Callout warning>

> **WARNING**: Setting `excludeVerboseContent` to `false` logs full queries and user parameters, which can contain sensitive data. Ensure your IAM policies limit access appropriately.

Reviewer: which IAM policy is this warning referring to? is it the role of the viewer?

Author: As sensitive information is logged on CW, we are advising the customers to limit who can read CW logs by scoping down IAM permissions to only the roles or users who truly need to see the logs, such as their DevOps or developers. Rephrased the warning.

</Callout>

## Configuration Properties

### `logging`
- `true`: Enables default logging.
- `DataLogConfig` object: Overrides one or more default fields.

### `DataLogConfig` Fields

1. **`retention?: LogRetention`**  
   - Number of days to keep the logs.  
   - Defaults to `'1 week'`.  
   - Supported values of [AppSync RetentionDays](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_logs.RetentionDays.html), such as:  
     `'1 day'`, `'3 days'`, `'5 days'`, `'1 week'`, `'2 weeks'`, `'1 month'`,  
     `'2 months'`, `'3 months'`, `'4 months'`, `'5 months'`, `'6 months'`,  
     `'1 year'`, `'13 months'`, `'18 months'`, `'2 years'`, `'5 years'`,  
     `'10 years'`, `'infinite'`

> 

Reviewer: does this line break and mix of ordered and unordered list break the rendering of the ordered list? if you inspect the resulting html it should ideally look something like
```
ol
  li
	  ul
	    li
	    li
	    li
  li
    ul
      li
			li
...
```

instead of 
```
ol
  li
    ul
      li
			li
ol start=2
  li
    ul
			li
			li
```

if it is messing with the rendering we could pivot to using `a`, `b`, `c` for the descriptions

Author: <img width="749" alt="Screenshot 2025-01-10 at 3 06 27 PM" src="https://github.com/user-attachments/assets/8c4cb4d6-0799-419e-bf9f-5faa2337d126" />
It looks like this currently.

## Default Logging

Default logging can be enabled by setting the `logging` property to `true` in the call to `defineData`. For example:

```ts title="amplify/data/resource.ts"
export const data = defineData({
  some_schema,
  some_authorizationModes,
  logging: true
});
```

Using `logging: true` applies the default configuration:
- `excludeVerboseContent: true`
- `fieldLogLevel: 'none'`
- `retention: '1 week'`

## Custom Logging

You can override individual fields by providing a `DataLogConfig` object. For example:

```ts title="amplify/data/resource.ts"
export const data = defineData({
  some_schema,
  some_authorizationModes,
  logging: {
    excludeVerboseContent: false,
    fieldLogLevel: 'all',
    retention: '1 month'
  }
});
```

<Callout warning>
**WARNING**: Setting `excludeVerboseContent` to `false` logs full queries and user parameters, which can contain sensitive data. Ensure your IAM policies limit access appropriately.
</Callout>

## Configuration Properties

### `logging`
- `true`: Enables default logging.
- `DataLogConfig` object: Overrides one or more default fields.

### `DataLogConfig` Fields

> 1. **`retention?: LogRetention`**  

Reviewer 2: `LogRetention` isn't defined in this doc. Link to its API reference definition (presumably, AppSync docs)

   - Number of days to keep the logs.  
   - Defaults to `'1 week'`.  
   - Possible values include:  
     `'1 day'`, `'3 days'`, `'5 days'`, `'1 week'`, `'2 weeks'`, `'1 month'`,  
     `'2 months'`, `'3 months'`, `'4 months'`, `'5 months'`, `'6 months'`,  
     `'1 year'`, `'13 months'`, `'18 months'`, `'2 years'`, `'5 years'`,  

>      `'10 years'`, `'infinite'`, etc.

Reviewer 2: nit: "etc" implies this is a freeform field, which is not the case

Suggest reword to something like:

```
Supported values of [AppSync `RetentionDays`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_logs.RetentionDays.html), such as `'1 day'`, ...
```

2. **`excludeVerboseContent?: boolean`**  
   - Defaults to `true`.  
   - When `false`, logs can contain the full query, response, and mapping templates.

> 3. **`fieldLogLevel?: DataLogLevel`**  

Reviewer 2: `DataLogLevel` isn't defined in this doc.  Link to its API reference definition (presumably, AppSync docs)
